### DISCUSSION TITLE
magick convert +append is very slow

### DISCUSSION THREAD

**Discussion by suteny0r** [ImageMagick/ImageMagick#5143] [Help]
Is there any way to optimize 
magick convert +append ?_0425.png output/B_0425.png 
Where the mask matches a left/right pair of files.
The left file is 4813x3058
The right file is 2413x3058
The append operation on  an Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz  takes about 20 seconds.
I have enabled magick's OpenCL support for my RTX-3090, but it does not reduce the time.
Here are a pair of sample files
(https://drive.google.com/drive/folders/13WHwnn2Msw9qM1YpmRztENtaDrRpZuBp?usp=sharing)

**Comment by fmw42** [comment]
I suggest you use proper Imagemagick syntax, thought it probably will not help. On Imagemagick 7, use magick, not magick convert. Also read the input images before doing any processing operation.

`magick left.png right.png +append +repage result.png`

With large image, depending upon your RAM availability, you may need to modify your policy.xml file.

See 

https://imagemagick.org/script/security-policy.php
https://imagemagick.org/script/resources.php
https://legacy.imagemagick.org/Usage/files/#massive

**Reply by suteny0r** [reply]
Using the syntax you provided, the operation time remained the same.

**Reply by suteny0r** [reply]
Pretty sure not running out of ram.  Only about half of my 16gb is allocated.
I used the -debug cache suggestion to see if operations were swapping part out to disk.  They were not.   Here is debug output in case there might be other clues. 

magick -debug cache L_0469.png R_0469.png +append +repage result.png
2022-05-10T15:56:24+00:00 0:00.040 0.000u 7.1.0 Cache magick[24012]: cache.c/OpenPixelCache/3793/Cache
  open [0] (Heap Memory, 1x1x3 12B)
2022-05-10T15:56:24+00:00 0:00.047 0.000u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy [0]
2022-05-10T15:56:24+00:00 0:00.057 0.000u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy
2022-05-10T15:56:24+00:00 0:00.062 0.000u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy
2022-05-10T15:56:24+00:00 0:00.069 0.000u 7.1.0 Cache magick[24012]: cache.c/OpenPixelCache/3793/Cache
  open L_0469.png[0] (Heap Memory, 4813x3058x3 168.436MiB)
2022-05-10T15:56:25+00:00 0:00.520 0.438u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy
2022-05-10T15:56:25+00:00 0:00.525 0.438u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy
2022-05-10T15:56:25+00:00 0:00.531 0.438u 7.1.0 Cache magick[24012]: cache.c/OpenPixelCache/3793/Cache
  open R_0469.png[0] (Heap Memory, 2413x3058x3 84.4454MiB)
2022-05-10T15:56:25+00:00 0:00.736 0.641u 7.1.0 Cache magick[24012]: cache.c/OpenPixelCache/3793/Cache
  open L_0469.png[0] (Heap Memory, 7226x3058x3 252.881MiB)
2022-05-10T15:56:25+00:00 0:01.120 2.500u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy L_0469.png[0]
2022-05-10T15:56:25+00:00 0:01.141 2.594u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy R_0469.png[0]
2022-05-10T15:56:25+00:00 0:01.409 4.234u 7.1.0 Cache magick[24012]: cache.c/OpenPixelCache/3793/Cache
  open result.png[0] (Heap Memory, 7226x3058x3 252.881MiB)
2022-05-10T15:56:46+00:00 0:21.853 24.656u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy result.png[0]
2022-05-10T15:56:46+00:00 0:21.881 24.672u 7.1.0 Cache magick[24012]: cache.c/DestroyPixelCache/1060/Cache
  destroy L_0469.png[0]

**Reply by snibgo** [reply]
Changing to the more current syntax doesn't affect the actual processing, so it won't affect the time. For that, see my post below.

**Reply by suteny0r** [reply]
I haven't modified policy.xml and I couldn't spot any directives that were active that might seem to constrain memory during read/write operations.

**Comment by snibgo** [comment]
You say:
> The append operation on an Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz takes about 20 seconds.

How do you know that? Perhaps the total command takes 20 seconds. How long does it take if you remove "+append"? For me, the time *increases*. Most time is needed for reading/writing the PNG files, not the append.

```
f:\web\im>timec %IMG7%magick L_0000.png R_0000.png +append out.png

TimeC C:\IM\ImageMagick-7.1.0-4-Q16-HDRI\magick L_0000.png R_0000.png +append ou
t.png
Seconds: 38.54

f:\web\im>timec %IMG7%magick L_0000.png R_0000.png  out.png

TimeC C:\IM\ImageMagick-7.1.0-4-Q16-HDRI\magick L_0000.png R_0000.png out.png
Seconds: 46.51
```
PNG files are massively slow. If we use TIFF instead:
```
TimeC C:\IM\ImageMagick-7.1.0-4-Q16-HDRI\magick L_0000.tiff R_0000.tiff +append
out.tiff
Seconds: 3.39
```

**Reply by suteny0r** [reply]
I think we have localized the problem to the PNG write routines.
if I pre-convert the two halves to .tiff   and then combine them output as .png  the time remains the same.
If I combine the two .tiff files and output as .tiff, the operation takes less than 1 second.
If I combine the two original .png files and output as .tiff the operation takes less than 1 second.
As long as ffmpeg is willing to assemble h.264 from .tiff frames, this will solve my problem

might want to have a developer look at the .png output routines.   :)    Thanks for making such an awesome product available.

**Comment by snibgo** [comment]
IM uses libpng to read and write PNG files. Any performance improvement would need to be in libpng.

**Comment by snibgo** [comment]
If this is the final image processing before combining into a video, you might consider using JPG format files.

**Reply by suteny0r** [reply]
ffmpeg ingested the tiff files without difficulty.  Are you suggesting that using the default settings, there maybe a qualitative advantage in using jpg over tiff as the intermediate image format?

I just noticed that the intermediate tiff files were much larger than the png files (7226x3058) png were 15mb, tiff were 68mb.   Frames of those dimensions at 24bit per pixel uncompressed works out to about 67mb, which might explain the speed difference.  The filesize of the PNG implies that it is using lossless compression, which slows things down.   Is there a selectable option for writing lossless, uncompressed png files?

**Reply by fmw42** [reply]
You could test using various TIFF compressions: ZIP, LZW, JPG or no compression.  PNG does not have an uncompressed mode that I know, but there are various quality values that you can try to see if one is faster.  See https://imagemagick.org/script/command-line-options.php#quality

**Reply by suteny0r** [reply]
For my use case, it's important for the intermediate format to be either losslessly compressed or uncompressed.   It's ok for intermediate size files to be very large.  It's not ok for it to be slow writing them, spending time compressing.   I haven't worked with png very much.  I've read that it has lossless mode, but didn't know about mandatory compression.  Sounds like tiff is a good choice for this process.  
The project is assembling multiple sections of video frames and encoding them into a video stream.  The resultant vide is less than 1% of the uncompressed frame sizes combined.

**Comment by snibgo** [comment]
>  Is there a selectable option for writing lossless, uncompressed png files?

No. PNG is always lossless compressed.

Best speed usually comes from a format internal to IM, MPC or MIFF. But most other software can't read those.

For best speed that ffmpeg can read, uncompressed TIFF is the obvious choice. Yes, these are large, but you can delete them after making the movie.

PNG is quite fast when used for what it is designed for: graphics, ie large areas of flat colours. For more photographic images, it is slow.

**Reply by suteny0r** [reply]
Here is more discussion of this issue:  
https://legacy.imagemagick.org/discourse-server~/viewtopic9e80.html?p=172440

I'm happy.  I can afford to give up the extra storage for speed.   The real lesson here is, PNG will always be relatively slow to write because it always compresses it's output.
In my case, producing a 15 second video, requiring appending two png files per frame, once combined, each frame is 66mb raw, 15mb png-compressed.
Having to append 470 pairs of frame images using png output took 2.5 hours.  Doing it with uncompressed tiff took 15 minutes.

Thanks again for the help.  Image Magick has no peer.

**Reply by Jmuccigr** [reply]
FWIW I use GraphicConverter on my mac for a lot of image work and its png output is a dog, too.
